### PR TITLE
Update build to use prebuilt extensions

### LINKED ISSUE #1077: Update Extensions to use federated build
Jupyterlab 3.0 has introduced a new way to build extensions (currently refereed to as federated builds) that leverage pip to make the extensions build dynamically. More information can be found in the docs https://jupyterlab.readthedocs.io/en/latest/user/extensions.html#extensions

We should update Elyra's extensions to use the new build system. Since this upgrade is non-trivial it was decided to not make the update part of Elyra's initial support of Jupyterlab 3.0

Some initial notes from when I had a working branch using federated builds:

- The new build setup require many new build files, many of which contain redundant metadata, we may want to find a way to generate the build files rather than have static files that need to be kept in sync.
- With the new build system, many of the build steps in our current Makefile may be redundant, in my previous "working" branch our Makefile seemed to rebuilding the code 3-4 times when leveraging the new build system.

### LINKED ISSUE #2339: start using prebuilt JupyterLab extensions (reduce install time)
Currently, we are using [source extensions](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#source-extensions) rather than [prebuilt extensions](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#prebuilt-extensions) for Jupyterlab, you can read about the differences [here](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#overview-of-extensions).

The gist of the problem is that it can take an __incredible__ amount of time to install Elyra right now, as you have to run `jupyter lab build`, and then restart JupyterLab. If we moved to prebuilt extensions, installing would be as simple as `pip install elyra`.

### PR DESCRIPTION
**Pull request by ajbozarth**
Updates Elyra to use prebuilt extensions, thus removing the `jupyter lab build` step from user installs.

Fixes #1077 
Fixes #2339
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ajbozarth** [comment]
I've moved this out of WIP but have left it as a draft. It is now fully function for local builds and watch mode. It is also Theoretically ready for releases, but that is untested and dependent on the sister PR in the template repo (PR linked above). I have not touched any of our Docker code yet and have not updated any docs, which is why this is still in draft

File: README.md [hunk 1]
```diff
@@ -114,6 +114,8 @@ Elyra can be installed from PyPI:
   conda install -c conda-forge "elyra[all]>=3.7.0"
   ```
 
+  Install Elyra from PyPI ( Elyra >= 3.1.0 ):
+
   ```bash
   pip3 install --upgrade "elyra[all]>=3.1.0" && jupyter lab build
   ```
```
Comment (kevin-bates) [inline README.md:121]: I'm finding this section of the README confusing and inconsistent.  In some places, the version in the "header" matches that in the command, and in others not.  In some places, `&& jupyter lab build` is appended, and in others not.

I'm assuming the lab build phrase is no longer necessary due to "prebuilt extensions", but it seems like some explanation to that effect is necessary.  Or remove all of the `&& jupyter lab build` and add a section titled: "When is a lab build necessary" or something related to Lab builds with an explanation as to why it's not required for elyra >= 3.7.

I suspect the version number in the "header" on L112 needs to reflect `3.7.0` as well.

I'm also not understanding the significance of `3.1.0` but do see the significance of `< 2`, although I think the "and above" comment in "JupyterLab 2.x is supported on Elyra 1.0.0 and above" should be updated to be "and below 3.0.0".

I know these aren't all necessarily related to prebuilt extensions, but it seems like a good time to clean things up.

Comment (ajbozarth) [reply README.md:121]: L112 is an error, I'll fix that. I'll also add a line explaining why `jupyter lab build` is needed for install prior to 3.7

As for the 3.1 difference, that was when we added the `[all]` to the install line.

File: docs/source/getting_started/installation.md [hunk 2]
```diff
@@ -51,41 +51,41 @@ Prior to version 3.1, the `elyra` package included all dependencies. Subsequent
 If you use `pip`, install Elyra with:
 
 ```bash
-pip3 install --upgrade elyra[all] && jupyter lab build
+pip3 install --upgrade elyra[all]
```
Comment (karlaspuldaro) [inline docs/source/getting_started/installation.md:54]: Just curious, do we still need to specifically add `[all]` to install all packages by default?

Comment (ajbozarth) [reply docs/source/getting_started/installation.md:54]: @akchinSTC @lresende you two know more about pip install details than I do

**Comment by ajbozarth** [comment]
Addressed CI test failures. The fix for code snippets was a "hack" and I opened up a followup https://github.com/elyra-ai/elyra/issues/2501 to properly address it.

File: docs/source/getting_started/installation.md (hunk 2 above)
Comment (kevin-bates) [reply docs/source/getting_started/installation.md:54]: > Just curious, do we still need to specifically add [all] to install all packages by default?

The `[all]` that is referenced here is really about how the [`extras_require` entry](https://github.com/elyra-ai/elyra/blob/43e6d9869c8e5bf2dd8213e8042c2d120d196fe7/setup.py#L110-L113) works wrt to pip.  In this case `[all]` is a "meta-entry" that encompasses the [4 separately installable "sub-packages"](https://github.com/elyra-ai/elyra/blob/43e6d9869c8e5bf2dd8213e8042c2d120d196fe7/setup.py#L43-L54), and not related to the changes in this PR.

File: build_requirements.txt [hunk 3]
```diff
@@ -1,2 +1,2 @@
-flake8>=3.5.0,<3.9.0
-flake8-import-order>=0.18.1
+jupyterlab>=3.0.17
+jupyter-packaging>=0.10,<2
```
Comment (kevin-bates) [inline build_requirements.txt:2]: Could we please remove this cap?
```suggestion
jupyter-packaging>=0.10
```

File: setup.py [hunk 4]
```diff
@@ -79,6 +74,7 @@
         'jsonschema>=3.2.0',
         'jupyter_core>=4.0,<5.0',
         'jupyter_client>=6.1.7',
+        'jupyter-packaging>=0.10,<2',
```
Comment (kevin-bates) [inline setup.py:77]: Is this a runtime dependency now?

Can we please remove the cap (per the previous comment)?

**Review by kevin-bates — COMMENTED** [review]
This looks good and is a very nice cleanup.  Just had a couple of comments relative to the `jupyter-packaging` dependency.

File: build_requirements.txt (hunk 3 above)
Comment (ajbozarth) [reply build_requirements.txt:2]: sure, I only left that in because the migration guide had it when I did the initial changes to setup.py

**Review by kevin-bates — APPROVED** [review]
LGTM - thanks Alex!  (EDIT: This may have been premature as I had only done a `make clean install` - so I've re-requested my review.)

**Comment by ajbozarth** [comment]
> Is this because I didn't run the "clean" script? Based on the message, it seems this could happen whenever a package dependency is updated.

Possibly, since how the artifacts is handled has changed, old files left in the previous extensions dir could cause issues. Running the clean-jupyterlab.sh script should remove those artifacts. For instance we used to link the extension libs(the ones listed in the ss), but now they are just included in the built extensions and lab doesn't need to know about them.

> While I suppose this may have always been the case, it seems like the git extension is producing output that I don't recall before...

IIRC git has always dumped a ton of junk into debug, it's a previously raised complaint on their repo

**Review by ptitzler — COMMENTED** [review]
`make elyra-image` fails:

```
#13 [7/9] RUN cd /tmp/elyra  && make install-all  && python3 -m pip install -r custom_requirements.txt  && rm custom_requirements.txt  && rm -rf /home/jovyan/.cache/yarn  && rm -rf /home/jovyan/.cache/Cypress  && rm -rf /tmp/elyra/node_modules  && rm -rf /opt/conda/share/jupyter/lab/staging/node_modules
#13 0.225 python -m pip install --upgrade pip
#13 0.601 Requirement already satisfied: pip in /opt/conda/lib/python3.9/site-packages (21.3.1)
#13 0.771 Collecting pip
#13 0.852   Downloading pip-22.0.3-py3-none-any.whl (2.1 MB)
#13 1.690 Installing collected packages: pip
#13 1.690   Attempting uninstall: pip
#13 1.690     Found existing installation: pip 21.3.1
#13 1.694     Uninstalling pip-21.3.1:
#13 1.938       Successfully uninstalled pip-21.3.1
#13 2.983 Successfully installed pip-22.0.3
#13 5.009 yarn install
#13 6.058 yarn install v1.22.17
#13 6.310 [1/4] Resolving packages...
#13 6.835 warning Resolution field "@lumino/widgets@1.17.0" is incompatible with requested version "@lumino/widgets@^1.19.0"
#13 6.886 warning Resolution field "@lumino/widgets@1.17.0" is incompatible with requested version "@lumino/widgets@^1.20.0"
#13 6.889 warning Resolution field "@lumino/widgets@1.17.0" is incompatible with requested version "@lumino/widgets@^1.31.1"
#13 7.393 [2/4] Fetching packages...
#13 54.44 warning mini-css-extract-plugin@0.5.0: Invalid bin field for "mini-css-extract-plugin".
#13 54.96 warning url-loader@1.1.2: Invalid bin field for "url-loader".
#13 103.6 error @npmcli/fs@1.1.0: The engine "node" is incompatible with this module. Expected version "^12.13.0 || ^14.15.0 || >=16". Got "12.4.0"
#13 103.6 error Found incompatible module.
#13 103.6 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
#13 103.7 make: *** [Makefile:125: yarn-install] Error 1
#13 ERROR: executor failed running [/bin/bash -o pipefail -c cd /tmp/elyra  && make install-all  && python3 -m pip install -r custom_requirements.txt  && rm custom_requirements.txt  && rm -rf $HOME/.cache/yarn  && rm -rf $HOME/.cache/Cypress  && rm -rf /tmp/elyra/node_modules  && rm -rf /opt/conda/share/jupyter/lab/staging/node_modules]: exit code: 2
------
 > [7/9] RUN cd /tmp/elyra  && make install-all  && python3 -m pip install -r custom_requirements.txt  && rm custom_requirements.txt  && rm -rf /home/jovyan/.cache/yarn  && rm -rf /home/jovyan/.cache/Cypress  && rm -rf /tmp/elyra/node_modules  && rm -rf /opt/conda/share/jupyter/lab/staging/node_modules:
#13 6.835 warning Resolution field "@lumino/widgets@1.17.0" is incompatible with requested version "@lumino/widgets@^1.19.0"
#13 6.886 warning Resolution field "@lumino/widgets@1.17.0" is incompatible with requested version "@lumino/widgets@^1.20.0"
#13 6.889 warning Resolution field "@lumino/widgets@1.17.0" is incompatible with requested version "@lumino/widgets@^1.31.1"
#13 7.393 [2/4] Fetching packages...
#13 54.44 warning mini-css-extract-plugin@0.5.0: Invalid bin field for "mini-css-extract-plugin".
#13 54.96 warning url-loader@1.1.2: Invalid bin field for "url-loader".
#13 103.6 error @npmcli/fs@1.1.0: The engine "node" is incompatible with this module. Expected version "^12.13.0 || ^14.15.0 || >=16". Got "12.4.0"
#13 103.6 error Found incompatible module.
#13 103.6 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
#13 103.7 make: *** [Makefile:125: yarn-install] Error 1
------
error: failed to solve: executor failed running [/bin/bash -o pipefail -c cd /tmp/elyra  && make install-all  && python3 -m pip install -r custom_requirements.txt  && rm custom_requirements.txt  && rm -rf $HOME/.cache/yarn  && rm -rf $HOME/.cache/Cypress  && rm -rf /tmp/elyra/node_modules  && rm -rf /opt/conda/share/jupyter/lab/staging/node_modules]: exit code: 2
make: *** [elyra-image] Error 1
```

**Review by ptitzler — COMMENTED** [review]
Note: `make elyra-image TAG=3.7.0` and `make kf-notebook-image TAG=3.7.0` (non-source code builds) cannot be validated until 3.7 was published

**Comment by kevin-bates** [comment]
After running `etc/scripts/clean-jupyterlab.sh` then `make clean install`, about 20 seconds after Lab presents itself in the browser, I get the following:
image

Do I need to do any additional "cleanup"?

> IIRC git has always dumped a ton of junk into debug, it's a previously raised complaint on their repo

Yeah, I still see the dumped git log.  I guess I hadn't noticed that before as it was probably lost in the other debug output (including all the LSP stuff!).

**Review by ptitzler — COMMENTED** [review]
The container image created by  `make kf-notebook` appears to work as expected.

File: Makefile [hunk 5]
```diff
@@ -166,18 +155,23 @@ install-gitlab-dependency:
 
 check-install:
 	jupyter serverextension list
-	jupyter server extension list
```
Comment (ptitzler) [inline Makefile:169]: Is this no longer required/useful? If it is not the command should also be removed from the validation instructions in the README and installation documentation.

Comment (kevin-bates) [reply Makefile:169]: I agree, but we need to keep `jupyter server extension list` (which comes from [`jupyter_server`](https://github.com/jupyter-server/jupyter_server/blob/62607ce28576e50358217534161f9e7c4aab2611/jupyter_server/extension/serverextension.py#L359)) and remove `jupyter serverextension list` (which comes from [`jupyter notebook`](https://github.com/jupyter/notebook/blob/26626343384195a1f4f5461ba42eb3e133655976/setup.py#L143)).

Comment (ajbozarth) [reply Makefile:169]: oh did I mix them up? I meant to delete the old redundant one. And I'll double check the docs

**Comment by ajbozarth** [comment]
I just pushed some update to the release script, but just like docker, further testing will have to wait till the PR is merged since it depends on git clone to pull the code to build

File: README.md [hunk 6]
```diff
@@ -148,33 +161,6 @@ Elyra can be installed from PyPI:
 
 Run the following commands to verify the installation. Note that in the example output below the `[version]` placeholder is displayed instead of an actual version identifier, which might change with every release.
 
-```bash
-jupyter serverextension list
```
Comment (kevin-bates) [inline README.md:152]: We should remove this same output from https://github.com/elyra-ai/elyra/blob/master/docs/source/getting_started/installation.md#verify-installation

Comment (ajbozarth) [reply README.md:152]: done

**Review by kevin-bates — APPROVED** [review]
LGTM - thanks Alex!

**Review by akchinSTC — COMMENTED** [review]
- confirmed `make elyra-image` fails but `make kf-notebook-image` succeeds in building(not tested in deployment)
- `make install-all` works and my usual dev pattern is all still functioning. Pipelines are submitting successfully.

**Review by akchinSTC — APPROVED** [review]

**Comment by marthacryan** [comment]
I only did a `make clean install` but it worked for me! Thanks!

**Pull request merged**